streaming/src/event_server.h

```diff
@@ -0,0 +1,229 @@
+#ifndef RAY_STREAMING_EVENT_SERVER_H
+#define RAY_STREAMING_EVENT_SERVER_H
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+#include <thread>
+#include <unordered_map>
+
+#include "channel.h"
+#include "ring_buffer.h"
+#include "util/streaming_util.h"
+
+namespace ray {
+namespace streaming {
+
+/// Data writer utilizes what's called an event-driven programming model instead
+/// loop forward polling driven that's employed in first simple version.
+/// Event driven model includes two important components: event server and event
+/// queue. In the process of data transmission, the inputer will first define
+/// the processing method of corresponding events. However, by triggering
+/// different events in actual operation, these events will be put into the event
+/// queue, and finally the event server will schedule the previously registered
+/// processing functions ordered by its priority.
+
+/// The event queue inherits from the universal ring queue. It mainly divides
+/// event into two different levels: normal event and urgent event, and the
+/// total size of the queue is the sum of them.
+template <class T>
+class EventQueue : public AbstractRingBufferImpl<T> {
```

Comment:
Does it make more sense to make `AbstractRingBufferImpl<T>` as a member variable of `EventQueue`?

I'm not sure ~

Reply:
EventQueue is a kind of ring buffer in our design.